### DISCUSSION TITLE
Detect .ply as binary

### DISCUSSION THREAD

**Discussion by Walther** [github-linguist/linguist#6974] [Classification] [answered]
[PLY (file format)](https://en.wikipedia.org/wiki/PLY_(file_format))
> PLY is a computer file format known as the Polygon File Format or the Stanford Triangle Format. It was principally designed to store three-dimensional data from 3D scanners. The data storage format supports a relatively simple description of a single object as a list of nominally flat polygons. A variety of properties can be stored, including color and transparency, surface normals, texture coordinates and data confidence values. The format permits one to have different properties for the front and back of a polygon.
> 
> There are two versions of the file format, one in ASCII, the other in binary.

Currently, there are [over 50 thousand .ply files](https://github.com/search?q=path%3A*.ply+NOT+is%3Afork&type=code) found on the GitHub code search. These are treated as code, and show up in statistics for lines added, pull request change sizes, and so on.

Even though the files can _technically_ be ASCII, they are almost universally generated by 3D modeling programs as output, used by 3D rendering programs as input, and easily go up to millions of lines in length.

Unfortunately, even the various workarounds proposed in the documentation and other relevant issues do not help. The following attempted settings in `.gitattributes` have no effect.

```
# attempt 1
*.ply binary
```

```
# attempt 2
*.ply binary
ply/** linguist-vendored
```

```
# attempt 3
ply/**/*.ply binary linguist-generated
```

```
# attempt 4
ply/**/*.ply -diff -merge -text linguist-generated
```

A pull request with some test models added shows up as 3 million lines of code added, potentially ruining any statistics and insights over time.

**Comment by lildude** [comment, marked as answer]
Unfortunately, there is no way of marking a file as binary and this has nothing to do with Linguist which is only used to detect the language; GitHub does not read the binary git attribute.

The closest you can get with Linguist is to mark the file as generated:

```gitattributes
*.ply linguist-generated
```

This will only suppress the content in diffs and won't count the files towards the language statistics.

That said, GitHub won't render the content of very large files in the diff by default anyway so the only benefit you'll get is the files won't be counted in the language stats. This won't however change the diff stats as these are based on the actual content of the file.

Interestingly Linguist doesn't even know about this language so it won't appear in the language stats of a repo nor get syntax highlighting. If you've got the time, we're happy to accept a PR that adds support for the language and marks the files as generated by default. This

**Reply by mk-pmb** [reply]
> Unfortunately, there is no way of marking a file as binary

I seem to be missing context. To what scope was that claim limited?
At least for git in general, the official method (chapter "Marking files as binary" in `man 5 gitattributes`) [was documented in 2011](https://github.com/git/git/commit/ab4356111afb28568bf1ad2e909b6ecf9f83a5e6), which is to use the `-diff` attribute in the `.gitattributes` file.
That part still is in the master branch today, so I assume we never updated the method.

> GitHub does not read the binary git attribute.

Probably same missing context: Why should GitHub check a "binary" attribute rather than "diff"?

Edit: Oh, I see. There is a built-in macro attribute "binary" that is equivalent to both `-diff` and `-text`. That may cause confusion depending on whether the parser used to read the `.gitattributes` file understands macros, whether it obeys user-confugured macros, and whether it knows the built-in macros.
And then for even more confusion, we have the `diff.*.binary` config option as an independent 3rd concept that overrides whatever would result from gitattributes.

**Comment by Walther** [comment]
Unfortunately, this does not seem to have any effect.
![gitattributes](https://github.com/user-attachments/assets/8c581f60-7dec-459c-a428-f3ac1f9ebac5)
![ply](https://github.com/user-attachments/assets/9cef4e71-57b5-4e6a-93b2-579c61428000)
<img width="300" alt="Screenshot 2024-07-30 at 12 17 06" src="https://github.com/user-attachments/assets/c99b9091-2a3c-4155-989b-b94735dc297d">


Compare to a glTF .bin file in another PR
![gltf-bin](https://github.com/user-attachments/assets/cd497817-0ac3-4dd7-8733-ac9c4b62f18b)
<img width="258" alt="Screenshot 2024-07-30 at 12 11 41" src="https://github.com/user-attachments/assets/078502e1-6d36-416e-ba9c-13f3b1b3aedb">

**Comment by lildude** [comment]
> Unfortunately, this does not seem to have any effect.

Indeed. I stated this...

> This won't however change the diff stats as these are based on the actual content of the file.
